File: backend/contacts.py
"""Contact channels stored in users.additional_contact (JSON list or legacy plain text)."""
from __future__ import annotations

import json
from typing import Any, Optional

CONTACT_TYPES = ("telegram", "whatsapp", "phone", "vk", "instagram", "other")

_LABELS = {
    "telegram": "Telegram",
    "whatsapp": "WhatsApp",
    "phone": "Телефон",
    "vk": "VK",
    "instagram": "Instagram",
    "other": "Другое",
}


def _normalize_type(raw: Any) -> str:
    value = str(raw or "").strip().lower()
    return value if value in CONTACT_TYPES else "other"


def _normalize_row(item: Any) -> Optional[dict]:
    if not isinstance(item, dict):
        return None
    value = str(item.get("value") or "").strip()
    if not value:
        return None
    return {"type": _normalize_type(item.get("type")), "value": value}


def parse_contacts(raw: Optional[str]) -> list[dict]:
    if raw is None:
        return []
    text = str(raw).strip()
    if not text:
        return []

    if text.startswith("["):
        try:
            data = json.loads(text)
        except (json.JSONDecodeError, TypeError):
            return [{"type": "other", "value": text}]
        if not isinstance(data, list):
            return [{"type": "other", "value": text}]
        rows: list[dict] = []
        for item in data:
            row = _normalize_row(item)
            if row:
                rows.append(row)
        return rows

    return [{"type": "other", "value": text}]


def serialize_contacts(contacts: Optional[list]) -> Optional[str]:
    if not contacts:
        return None
    rows: list[dict] = []
    for item in contacts:
        row = _normalize_row(item)
        if row:
            rows.append(row)
    if not rows:
        return None
    return json.dumps(rows, ensure_ascii=False)


def format_contacts(contacts: Optional[list]) -> str:
    if not contacts:
        return ""
    parts: list[str] = []
    for item in contacts:
        row = _normalize_row(item)
        if not row:
            continue
        label = _LABELS.get(row["type"], "Другое")
        parts.append(f"{label}: {row['value']}")
    return " · ".join(parts)

File: src/utils/contacts.js
/** Client helpers for profile contact channels (mirrors backend/contacts.py). */

export const CONTACT_TYPES = [
  { value: 'telegram', label: 'Telegram', placeholder: '@username' },
  { value: 'whatsapp', label: 'WhatsApp', placeholder: '+79991234567' },
  { value: 'phone', label: 'Телефон', placeholder: '+79991234567' },
  { value: 'vk', label: 'VK', placeholder: 'vk.com/id…' },
  { value: 'instagram', label: 'Instagram', placeholder: '@username' },
  { value: 'other', label: 'Другое', placeholder: 'Как связаться' },
];

const LABELS = Object.fromEntries(CONTACT_TYPES.map((t) => [t.value, t.label]));

export function parseContacts(raw) {
  if (raw == null) return [];
  const text = String(raw).trim();
  if (!text) return [];

  if (text.startsWith('[')) {
    try {
      const data = JSON.parse(text);
      if (!Array.isArray(data)) return [{ type: 'other', value: text }];
      return data
        .map((item) => {
          if (!item || typeof item !== 'object') return null;
          const value = String(item.value || '').trim();
          if (!value) return null;
          const type = CONTACT_TYPES.some((t) => t.value === item.type) ? item.type : 'other';
          return { type, value };
        })
        .filter(Boolean);
    } catch {
      return [{ type: 'other', value: text }];
    }
  }

  return [{ type: 'other', value: text }];
}

export function contactsFromUser(user) {
  if (!user) return [];
  if (Array.isArray(user.contacts) && user.contacts.length) {
    return user.contacts.map((c) => ({
      type: CONTACT_TYPES.some((t) => t.value === c.type) ? c.type : 'other',
      value: String(c.value || '').trim(),
    })).filter((c) => c.value);
  }
  return parseContacts(user.additional_contact);
}

export function formatContacts(contacts) {
  if (!Array.isArray(contacts) || !contacts.length) return '';
  return contacts
    .map((c) => {
      const value = String(c?.value || '').trim();
      if (!value) return null;
      const type = CONTACT_TYPES.some((t) => t.value === c.type) ? c.type : 'other';
      return `${LABELS[type]}: ${value}`;
    })
    .filter(Boolean)
    .join(' · ');
}

export function emptyContactRow() {
  return { type: 'telegram', value: '', key: `${Date.now()}-${Math.random().toString(36).slice(2, 7)}` };
}
